---
title: "Data Preparation Workflow (Section 1)"
author: "Siyuan"
date: "7/25/2020"
output: 
  html_document:
    fig_width: 10
    fig_height: 8
---

```{r, include = FALSE}
knitr::opts_chunk$set(echo = TRUE,    # show code
                      message = FALSE # hide console output
                      )
```

This is a brief introduction of how to use this library to collect and preprocess data that would be used for CellScore analysis. In this tutorial, we demonstrate how to use this package to download the data, summarize gene-level count data from transcript-level count, normalize count data and plot the data for PCA analysis.


In this section, we start from collecting the selected data from DEE2

### Step 1: Load the dependencies
```{r setup}
library(DEE2HsapienData)
library(SummarizedExperiment)
```

### Step 2: Download the selected hsapien data and compose a 'SummarizedExperiment' object
First, We create a 'temp' folder and run 'GetDEE2Data' function to:

1. Download the data (a list object) from DEE2.
2. Compose a SummarizedExperiment object named 'sExpr' and add it to the list object.

Then we save the list object in the 'temp' folder.
```{r}
## Creat a 'temp' directory in current working directory if it is not created
## already. In this tutorial, we use it to save the data. 
tempDirPath <- paste(getwd(), 'temp', sep = '/')
dir.create(tempDirPath)

## Get the selected homosapien data that is summarized in a 'List' Object.
## Internally, 'GetDEE2Data' uses 'rio' library to import pheno data, which by 
## default supports a wide range of data formats including tsv, csv, xls and many
## more. You could check the following link to find out which formats are supported
## by default: https://github.com/leeper/rio.
##
## In this example, the pheno data is recorded in an '.ods' file, which is
## not supported by default, so we have to explicitly load 'readODS' library to
## support the IO. 
library(readODS)
dee2Data <- GetDEE2Data(
  sampleInfoPath = system.file("extdata", "SelectedDEE2_v2.ods",
             package="DEE2HsapienData"),
  'hsapiens', 
  counts = 'GeneCounts')
```
### Step 3: Summarize the transcript-level DEE2 data to the gene-level count data using 'tximport'
When we download the data from DEE2, the data includes both the gene- and 
transcript-level count data. The gene-level count data from DEE2 was summarized with 'STAR' aligner. We are also interested in summarizing the transcript-level count data with 'tximport' to gene-level count and compare it with gene-level data downloaded from DEE2. The following code summarizes transcript-level DEE2 data to the gene-level count data using 'tximport' and add it as an element in the 'dee2Data' list object.

### Step 4: Generate 2D and 3D PCA plots for raw data and save the data  
We plot the gene-level count data that is downloaded from DEE2 and summarized from transcript-level data respectively. 
```{r}
PcaPlot(dee2Data)
```
```{r}
PcaPlot2d <- PcaPlot(dee2Data, plot = '2d');
PcaPlot2d[[1]]
PcaPlot2d[[2]]
PcaPlot2d[[3]]
```
```{r}
TsnePlot(dee2Data)
```
Save the downloaded and preprocessed data in the 'temp' folder.

```{r}
saveRDS(dee2Data, file=paste(tempDirPath, 'dee2Data', sep='/'))
```
